### PR TITLE
Don't use hash to check if binding worked

### PR DESCRIPTION
**Pull request by CarlSchwan**
using password_hash is expensive and should be used for hashing
passwords when saving them in the database. Here we just want to see if
the bind was already done with the given password.

Time spent on testing.nextcloud.com for password hashing 160ms:

image

### CODE REVIEW

**Review by come-nc — CHANGES_REQUESTED** [review]
Maybe name it sum instead of hash and use a fast checksum method like sha instead?
I dislike storing the clear password in a random variable, it might cause problems down the road.

**Comment by come-nc** [comment]
> Maybe name it sum instead of hash and use a fast checksum method like sha instead? I dislike storing the clear password in a random variable, it might cause problems down the road.

This sum could be of dn+prefix+pwd to test if the bind was with the same parameters in one comparison.

**Comment by CarlSchwan** [comment]
> Maybe name it sum instead of hash and use a fast checksum method like sha instead? I dislike storing the clear password in a random variable, it might cause problems down the road.

I switched to an md5 hash instead. It's still quite useless in my opinion since the password is already accessible in clear from the app config database, so adding a hash to compare two strings is not required even if the string is confidential.

At least with md5, the hashing will be quite a bit faster.

**Comment by come-nc** [comment]
> I switched to an md5 hash instead. It's still quite useless in my opinion since the password is already accessible in clear from the app config database, so adding a hash to compare two strings is not required even if the string is confidential.

Yeah I know we store the data but the idea is to not have it in too many variables to not encourage its use everywhere, and be able to track each use by looking at configuration reads.

File: apps/user_ldap/lib/Connection.php [hunk 1]
```diff
@@ -690,7 +687,7 @@ public function bind() {
 
 		$this->bindResult = [
 			'dn' => $this->configuration->ldapAgentName,
-			'hash' => \OC::$server->getHasher()->hash($this->configPrefix . $this->configuration->ldapAgentPassword),
+			'hash' => md5($this->configPrefix . $this->configuration->ldapAgentPassword),
```
Comment (come-nc) [inline apps/user_ldap/lib/Connection.php:690]:
```php
￼			'sum' => md5($this->configuration->ldapAgentName . $this->configPrefix . $this->configuration->ldapAgentPassword),
```

File: apps/user_ldap/lib/Connection.php [hunk 2]
```diff
@@ -673,10 +673,7 @@ public function bind() {
 		if (
 			count($this->bindResult) !== 0
 			&& $this->bindResult['dn'] === $this->configuration->ldapAgentName
-			&& \OC::$server->getHasher()->verify(
-				$this->configPrefix . $this->configuration->ldapAgentPassword,
-				$this->bindResult['hash']
-			)
+			&& md5($this->configPrefix . $this->configuration->ldapAgentPassword) === $this->bindResult['hash']
```
Comment (come-nc) [inline apps/user_ldap/lib/Connection.php:676]:
```php
			&& md5($this->configuration->ldapAgentName . $this->configPrefix . $this->configuration->ldapAgentPassword) === $this->bindResult['sum']
```
and remove the dn comparison.
You can also replace `count($this->bindResult) !== 0` by `isset($this->bindResult['sum'])` to please psalm.

File: apps/user_ldap/lib/Connection.php (hunk 1 above)
Comment (come-nc) [reply apps/user_ldap/lib/Connection.php:690]: (and remove the dn and hash keys)

**Review by come-nc — APPROVED** [review]

**Review by PVince81 — APPROVED** [review]
:+1:

**Comment by blizzz** [comment]
sha256 over md5 to be more robust against collisions

**Pull request merged**